src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/PooledStringBuilder.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Text;
+
+namespace System.Reflection.Internal
+{
+    /// <summary>
+    /// The usage is:
+    ///        var inst = PooledStringBuilder.GetInstance();
+    ///        var sb = inst.builder;
+    ///        ... Do Stuff...
+    ///        ... sb.ToString() ...
+    ///        inst.Free();
+    /// </summary>
+    internal class PooledStringBuilder
+    {
+        public readonly StringBuilder Builder = new StringBuilder();
+        private readonly ObjectPool<PooledStringBuilder> _pool;
+
+        private PooledStringBuilder(ObjectPool<PooledStringBuilder> pool)
+        {
+            Debug.Assert(pool != null);
+            _pool = pool;
+        }
+
+        public int Length
+        {
+            get { return this.Builder.Length; }
+        }
+
+        public void Free()
+        {
+            var builder = this.Builder;
+
+            // do not store builders that are too large.
+            if (builder.Capacity <= 1024)
+            {
+                builder.Clear();
+                _pool.Free(this);
+            }
+        }
+
+        [Obsolete("Consider calling ToStringAndFree instead.")]
```

Comment:
We don't need this internal obsolete method. 

Also, should we switch to StringBuilderCache in Common\?